pkg/ipamd/datastore/data_store.go

```diff
@@ -443,207 +445,113 @@ func (ds *DataStore) writeBackingStoreUnsafe() error {
-		if err := ds.writeBackingStoreUnsafe(); err != nil {
-			ds.log.Warnf("Unable to update backing store: %v", err)
-			// Continuing because 'force'
-		}
+	curENI.AvailableIPv4Cidrs[strIPv4Cidr] = &CidrInfo{
+		Cidr:          ipv4Cidr,
+		IPv4Addresses: make(map[string]*AddressInfo),
+		IsPrefix:      isPrefix,
 	}
-
-	ds.total--
-	// Prometheus gauge
-	totalIPs.Set(float64(ds.total))
-
-	delete(curENI.IPv4Addresses, ipv4)
-
-	ds.log.Infof("Deleted ENI(%s)'s IP %s from datastore", eniID, ipv4)
+	ds.log.Infof("Added ENI(%s)'s IP/Prefix %s to datastore", eniID, strIPv4Cidr)
 	return nil
 }
 
-// DelIPv4AddressFromStore delete an IP of ENI from datastore
-func (ds *DataStore) DelIPv4PrefixFromStore(eniID string, ipv4Prefix string, force bool) error {
+func (ds *DataStore) DelIPv4CidrFromStore(eniID string, cidr net.IPNet, force bool) error {
 	ds.lock.Lock()
 	defer ds.lock.Unlock()
 
 	curENI, ok := ds.eniPool[eniID]
 	if !ok {
+		ds.log.Debugf("Unknown ENI %s while deleting the CIDR", eniID)
 		return errors.New(UnknownENIError)
 	}
+	strIPv4Cidr := cidr.String()
+	ipv4 := cidr.IP.String()
 
-	//split IP here
-	ip, ipnet, err := net.ParseCIDR(ipv4Prefix)
-	if err != nil {
-		return err
-	}
-
-	ipv4 := ip.String()
-	prefixlen := ipnet.Mask
-	ds.log.Debugf("Adding IP %s and prefix len %s", ipv4, prefixlen)
-
-	ipPrefix, ok := curENI.IPv4Prefixes[ipv4]
+	_, ok = curENI.AvailableIPv4Cidrs[strIPv4Cidr]
 	if !ok {
+		ds.log.Debugf("Unknown %s CIDR", strIPv4Cidr)
 		return errors.New(UnknownIPError)
 	}
 
-	if ipPrefix.UsedIPs != 0 {
-		if !force {
-			return errors.New(IPInUseError)
-		}
-		ds.log.Warnf("Force deleting assigned Prefix %s on eni %s", ipv4, eniID)
-		forceRemovedIPs.Inc()
-		//Cleanup datastore /32 ips
-		prefixDB := ipPrefix.AllocatedIPs
+	deletableCidr := curENI.AvailableIPv4Cidrs[strIPv4Cidr]
 
-		for strPrivateIPv4, _ := range prefixDB.UsedIPs {
-			addr := curENI.IPv4Addresses[strPrivateIPv4]
+	// SIP case : This runs just once
+	// PD case : if (force is false) then if there are any unassigned IPs, those will get freed but the first assigned IP will
+	//  break the loop, should be fine since freed IPs will be reused for new pods.
+	updateBackingStore := false
+	for _, addr := range deletableCidr.IPv4Addresses {
+		if addr.Assigned() {
+			if !force {
+				return errors.New(IPInUseError)
+			}
+			forceRemovedIPs.Inc()
 			ds.unassignPodIPv4AddressUnsafe(addr)
+			updateBackingStore = true
 		}
+	}
+	if updateBackingStore {
 		if err := ds.writeBackingStoreUnsafe(); err != nil {
 			ds.log.Warnf("Unable to update backing store: %v", err)
 			// Continuing because 'force'
 		}
 	}
-
-	_, numIPsPerPrefix, _ := GetPrefixDelegationDefaults()
-	ds.total = ds.total - numIPsPerPrefix
-	ds.allocatedPrefix--
-	// Prometheus gauge
+	ones, bits := cidr.Mask.Size()
+	ds.total -= 1 << (bits - ones)
+	if deletableCidr.IsPrefix {
+		ds.allocatedPrefix--
+	}
 	totalIPs.Set(float64(ds.total))
+	delete(deletableCidr.IPv4Addresses, ipv4)
```

Comment:
why this is needed.

Reply:
Yes, this is not needed since we free up the CIDR block in the next line.